subsys/dfu/dfu_target/src/dfu_target_mcuboot.c

```diff
@@ -174,6 +174,35 @@ int dfu_target_mcuboot_write(const void *const buf, size_t len)
 	return dfu_target_stream_write(buf, len);
 }
 
+
+#define _DO_ERROR	0
+#define _DO_FLATTTEN 1
+#define _DO_ERASE	2
+
+/**
+ * Check what way of target done should be used.
+ */
+static int _chose_done_way(void)
+{
+#if IS_ENABLED(CONFIG_FLASH_HAS_EXPLICIT_ERASE)
+#if IS_ENABLED(CONFIG_FLASH_HAS_NO_EXPLICIT_ERASE)
+	const struct flash_parameters *fparams = flash_get_parameters(secondary_dev[cur_sec_img]);
+
+	/* Stream flash does not rely on erase, it does it when device needs it */
+	if (!(flash_params_get_erase_cap(fparams) & FLASH_ERASE_C_EXPLICIT)) {
+		return _DO_FLATTEN;
+	}
+#endif
```

Comment:
You do not have to do this this way, I know that stream does but that may need fixing.
If you look at the `flash_params_get_erase_cap` it has the same logic inside, so if you do not care about wasted `flash_get_parameters` call, the `flash_params_get_erase_cap` will return 0 if there is not device with erase, check it against  `FLASH_ERASE_C_EXPLICIT`.